Add DuckDuckGo Bangs inspired search-plus extension

franzwilhelm opened the pull request:
## Description

Search Plus implements DuckDuckGo-style bangs search directly in Raycast without sending traffic through third-party websites. It allows users to search specific websites using shortcuts like:

- `!g cats` to search Google for "cats"
- `!w cats` to search Wikipedia for "cats" 
- `Help me fix my code !t3` to use t3.chat for AI code help

For the best experience, users can set up Search Plus as a Fallback Command in Raycast, enabling bang searches directly from the main Raycast search.

Made this[ based on inspiration from Theo Browne on YouTube](https://www.youtube.com/watch?v=_DnNzRaBWUU), and I want his take on the PR if he has the time to review it.

## Screencast


### Using the search command
https://github.com/user-attachments/assets/9cf680ea-a841-4450-aaba-371fba3b84c5

### Browsing available search engines
https://github.com/user-attachments/assets/9cb6341f-2244-44e1-bd3e-9abe414e6518


## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

franzwilhelm commented:
People in the chat of the youtube video are saying a good name would be search router. I'm open for name suggestions here

![image](https://github.com/user-attachments/assets/6c07920a-7a64-4945-a788-e1a1105ad6e4)

franzwilhelm commented:
@lacherogwu I tried to add you as a contributor to the project, but your Raycast user was not found:

```
Not found (https://www.raycast.com/api/v1/users/lacherogwu RequestID: ac9c86f6-7ac5-4ec0-b3af-09ef78d6e702)
```

@pernielsentikaer: @lacherogwu should be a contributor as he had some great suggestions in his PR published some time after this one.

pernielsentikaer commented:
@lacherogwu what's your Raycast handle?

pernielsentikaer commented:
This seems to be the same as #17297. Could we work together on creating one by taking the best of both?

lacherogwu commented:
@pernielsentikaer 
my raycast handle is [lachero](https://www.raycast.com/lachero)

Perhaps you could also take a look at my submission, #17436, which I believe offers a more elegant and well-structured solution.

franzwilhelm commented:
@pernielsentikaer: As pointed out, it looks like there are now four PRs addressing the same problem. Here's a quick summary for clarity:

## Quicklink extension wrappers around `unduck.link`:
- https://github.com/raycast/extensions/pull/17297
- https://github.com/raycast/extensions/pull/17406

These PRs wrap a quicklink with `https://unduck.link?q={query}`. Unduck.link itself is a web-based implementation of [DuckDuckGo Bangs!](https://duckduckgo.com/bangs). While this approach works, I believe adding extensions for every link on the web could lead to unnecessary bloat in the Raycast store.

## Local-first [DuckDuckGo Bangs!](https://duckduckgo.com/bangs) implementations:
- https://github.com/raycast/extensions/pull/17418 (This PR, submitted Feb 27th)
- https://github.com/raycast/extensions/pull/17436 (submitted Feb 28th)

These PRs implement DuckDuckGo Bangs! functionality (similar to unduck.link) directly within Raycast, allowing users to bypass the need for a proxy URL. 

Regarding the overlap between this PR and @lacherogwu's, it seems we both independently arrived at similar solutions around the same time. This PR was submitted about 15 hours earlier, but I noticed some great implementation details in @lacherogwu's PR, which I’ve incorporated here. I see this as an opportunity for collaboration, as we’re both working toward the same goal: providing users with the best possible extension.

This PR also includes additional functionality that I believe enhances the user experience, such as the ability to browse available bangs, try them out, and set a default directly within the UI. As @lacherogwu pointed out, I opted not to use the Raycast preferences API in this implementation, as I found it provided a less intuitive experience compared to letting users select options during their search. However, I’m open to feedback and happy to adjust this approach if needed—that’s why this PR is still under review.

Looking forward to hearing everyone’s thoughts and working together to deliver the best solution for users!

franzwilhelm commented on extensions/search-plus/scripts/update-db.ts:
```diff
@@ -0,0 +1,144 @@
+import fs from "fs";
+import path from "path";
+
+export const DB_DIR = "./assets";
+export const DB_PATH = path.join(DB_DIR, "search-plus.db");
```
Not possible to use `@raycast/api` when not running in the Raycast environment. Assets path will never change

franzwilhelm commented on extensions/search-plus/src/search.tsx:
```diff
@@ -0,0 +1,51 @@
+type Props = LaunchProps<{ arguments: { query: string }; fallbackText?: string }>;
+
+export default async function search(props: Props) {
+  const query = (props.arguments.query ?? props.fallbackText) as string;
```
Not possible for both to be empty

franzwilhelm commented:
# Update

- [x] Addressed issues detected by automated PR reviews
- [x] Moved over to setting default search using preferences API as noticed by @lacherogwu 
- [x] Added  @lacherogwu (lachero) as contributor to package.json

franzwilhelm commented:
@lacherogwu @jafupy @andrewbellucci Do you think the extension should be named unduck like Theo's implementation, or should we stay closer to the "bangs" concept in the naming? The reason I'm asking is that DuckDuck Go had the original implementation, and we should choose the naming that leads to more people discovering the extension in the store. It might be called something else, but having unduck as one of the keywords in search. Let's discuss and land on something together?

pernielsentikaer commented:
Following replies from third parties

lacherogwu commented:
@franzwilhelm I think the name doesn't matter too much in terms of discoverability. Personally, the idea came from Theo, and I really appreciate the work he did with the [unduck.link](https://unduck.link/) webpage. I believe it would be fitting to stick with the name "Unduck," as it carries the meaning and intent behind Theo's original implementation, which I also found appealing.

pernielsentikaer commented:
It is certain that having four extensions performing the same functions is not feasible; it would be preferable to consolidate all features and ideas into a single extension.

franzwilhelm commented:
@lacherogwu Given how many people in the comments on YouTube liked the name "Search Router", I changed the name of the extension to that, and added "unduck" as a keyword. This made `npm run publish` make a new PR: https://github.com/raycast/extensions/pull/17518

Therefore I'm suggesting closing this one in favor of the newly created one

lacherogwu commented:
@franzwilhelm I've got nothing to add to the other points raised. My only concern is that the usage of SQLite here seems unnecessary. The file is only 2.5MB with simple objects - array.find() will outperform SQLite at this scale because:
1. In-memory operations avoid disk I/O overhead
2. Modern JS engines optimize array operations well
3. For small datasets like this, the simplicity of array methods provides better performance-to-complexity ratio

If you think the page view command is necessary, why not push it to my extension instead?

franzwilhelm commented:
@lacherogwu Thank you for your guidance! 

I initially struggled with JS Heap out of memory issues, as mentioned earlier. After considering your feedback on the local object approach, I revisited my original array implementation and identified the root cause: a base case that was attempting to display all bangs during search.

I've successfully resolved this issue and reverted to the array approach, which addresses your valid concerns:

![Memory issue resolved](https://github.com/user-attachments/assets/da642c66-26da-4185-95f0-09a67e975f19)

Since your feedback was crucial in this solution, I've added you as a co-author to this commit:
https://github.com/raycast/extensions/pull/17518/commits/48a21db479968c85fdc661f684ebe03d29d61508

I believe we now have the best of both worlds in the new PR here: https://github.com/raycast/extensions/pull/17518

franzwilhelm commented:
@pernielsentikaer This means we've now together resolved the concerns @lacherogwu had with my implementation. 

# The new PR is ready for review here: https://github.com/raycast/extensions/pull/17518
(new PR because ray publish created a new PR based on the name change)

pernielsentikaer commented:
Closing in favor of https://github.com/raycast/extensions/pull/17518
